src/frequenz/pymdownx/superfences/filter_lines/__init__.py

```diff
@@ -0,0 +1,272 @@
+
+        - `start`
+        - `start:end`
+        - `:end`
+        - `start:`
+
+        where `start` and `end` are integers.
+
+        If `start` is empty, it is assumed to be 1.
+        If `end` is empty, it is assumed to be the end of the file.
+        If `start` appears without `:`, it is assumed to be both `start` and `end`.
+
+        Lines are 1-indexed.
+
+        Args:
+            text: String to parse.
+
+        Returns:
+            The created range.
+
+        Raises:
+            ValueError: If the string is invalid.
+        """
+        splitted = text.split(":", 1)
+        stripped = map(lambda s: s.strip(), splitted)
+        match tuple(map(lambda s: s if s else None, stripped)):
+            case (None,):
+                raise ValueError("Empty start")
+            case () | (None, None):
+                raise ValueError("Both start and end are empty")
+            case (str() as start,):
+                return cls(start=int(start), end=int(start))
+            case (str() as start, None):
+                return cls(start=int(start))
+            case (None, str() as end):
+                return cls(end=int(end))
+            case (str() as start, str() as end):
+                return cls(start=int(start), end=int(end))
+            case _ as invalid:
+                raise ValueError(f"Invalid: {invalid!r}")
+
+    def __str__(self) -> str:
+        """Get the string representation."""
+        if self.start is None:
+            assert self.end is not None
+            return f":{self.end}"
+        if self.end is None:
+            return f"{self.start}:"
+        return f"{self.start}:{self.end}"
+
+
+@dataclasses.dataclass(frozen=True)
+class LinesRanges:
+    """A set of line ranges."""
+
+    ranges: Set[LinesRange]
+    """The lines ranges."""
+
+    def __post_init__(self) -> None:
+        """Validate."""
+        if not self.ranges:
+            raise ValueError("Cannot have empty ranges")
+
+    def __contains__(self, item: int) -> bool:
+        """Whether the item is inside any of the ranges."""
+        return any(item in r for r in self.ranges)
+
+    @classmethod
+    def parse(cls, text: str) -> tuple[Self | None, list[ValueError]]:
+        """Create from a string.
+
+        The string must be a comma-separated list of ranges, where each range is in the
+        format described in
+        [`LinesRange.parse`][frequenz.pymdownx.superfences.filter_lines.LinesRange.parse].
+
+        If no ranges are given, `None` is returned. If any ranges are invalid, they are
+        ignored and a list of errors is returned.
+
+        Args:
+            text: String to parse.
+
+        Returns:
+            The created ranges, or `None` if no ranges are given, plus a list of errors,
+                if any.
+        """
+        ranges: set[LinesRange] = set()
+        errors: list[ValueError] = []
+        for n, r in enumerate(text.split(","), start=1):
```

Comment:
maybe use `range` or `range_str` here instead of `r`?

Reply:
Updated